Code/Framework/AzToolsFramework/AzToolsFramework/Viewport/SharedViewBookmarkComponent.cpp

```diff
@@ -0,0 +1,116 @@
+                    ->Attribute(AZ::Edit::Attributes::AutoExpand, true)
+                    ->Attribute(AZ::Edit::Attributes::IndexedChildNameLabelOverride, &EditorViewBookmarks::GetBookmarkLabel);
+            }
+        }
+    }
+
+    AZStd::string EditorViewBookmarks::GetBookmarkLabel(int index) const
+    {
+        return AZStd::string::format("View Bookmark %d", index);
+    }
+
+    void SharedViewBookmarkComponent::Reflect(AZ::ReflectContext* context)
+    {
+        EditorViewBookmarks::Reflect(context);
+
+        if (auto* serializeContext = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serializeContext->RegisterGenericType<EditorViewBookmarks>();
+
+            serializeContext->Class<SharedViewBookmarkComponent, EditorComponentBase>()->Version(0)->Field(
+                "ViewBookmarks", &SharedViewBookmarkComponent::m_viewBookmark);
+
+            if (AZ::EditContext* editContext = serializeContext->GetEditContext())
+            {
+                editContext
+                    ->Class<SharedViewBookmarkComponent>(
+                        "Shared View Bookmark Component", "The ViewBookmark Component allows to store bookmarks for a prefab")
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                    ->Attribute(AZ::Edit::Attributes::AddableByUser, true)
+                    ->Attribute(AZ::Edit::Attributes::AppearsInAddComponentMenu, AZ_CRC("Level", 0x9aeacc13))
+                    ->Attribute(AZ::Edit::Attributes::Category, "View Bookmarks")
+                    ->Attribute(AZ::Edit::Attributes::Icon, "Icons/Components/Comment.svg")
+                    ->Attribute(AZ::Edit::Attributes::ViewportIcon, "Icons/Components/Viewport/Comment.svg")
+                    ->Attribute(AZ::Edit::Attributes::AutoExpand, true)
+                    ->DataElement(
+                        AZ::Edit::UIHandlers::Default, &SharedViewBookmarkComponent::m_viewBookmark, "ViewBookmarks", "ViewBookmarks")
+                    ->Attribute(AZ::Edit::Attributes::AutoExpand, false);
+            }
+        }
+    }
+
+    AZStd::optional<ViewBookmark> SharedViewBookmarkComponent::GetBookmarkAtIndex(int index) const
+    {
+        if (index >= 0 && index < m_viewBookmark.m_viewBookmarks.size())
+        {
+            return AZStd::optional<ViewBookmark>(m_viewBookmark.m_viewBookmarks[index]);
+        }
+        return AZStd::optional<ViewBookmark>();
+    }
+
+    void SharedViewBookmarkComponent::AddBookmark(ViewBookmark viewBookmark)
+    {
+        m_viewBookmark.m_viewBookmarks.push_back(viewBookmark);
+    }
+
+    bool SharedViewBookmarkComponent::RemoveBookmarkAtIndex(int index)
+    {
+        if (index >= 0 && index < m_viewBookmark.m_viewBookmarks.size())
+        {
+            m_viewBookmark.m_viewBookmarks.erase(m_viewBookmark.m_viewBookmarks.cbegin() + index);
+            return true;
+        }
+
+        return false;
+    }
```

Comment:
Since `m_viewBookmarks` is a `AZStd::vector`, after removing a bookmark will cause that all the indices that other users might have kept will be invalid. This is an important remark that must be commented in the interface, because there are other functions that use index as well.

If index (or id) for other elements must remain valid after calling `RemoveBookmarkAtIndex` then another data structure needs to be used, maybe a `unordered_map<index, ViewBookmark>`.

Reply:
A regular sorted `AZStd::map` might be a better data structure here, if in-order iteration is needed.

Reply:
Any comment on this @AMZN-Igarri ?

Reply:
yes I agree, I think I might go for `AZStd::map` since in-order iterations is going to be needed